### ISSUE TITLE
Information loss for Dates outside of the api

### ISSUE THREAD

**Issue by leonidas-o** [vapor/vapor#1977]
I'm not sure if this is a pure fluent issue or one the helper methods inside vapor.
The iso8601 format is a kind of broadly used default format. It is incorporated and has helpers like `jsonDecoder.dateDecodingStrategy = .iso8601`, `formatter.calendar = Calendar(identifier: .iso8601)`, in iOS and MacOs there is even a `ISO8601DateFormatter()`. It is recommended for working `internally` with dates.
Right now if you add createdAt/updatedAt properties to your model:
```
extension MyModel {
    static var createdAtKey: TimestampKey? = \.createdAt
    static var updatedAtKey: TimestampKey? = \.updatedAt
}
```
It will be saved with milliseconds `2019-05-14 19:54:06.91973`.
The iso8601 representation has no milliseconds, so it is `2019-05-14T19:54:06Z`. Therefore as soon as your api response contains dates, you loose the milliseconds information.

The `updatedAt`/`createdAt` date is saved with milliseconds (2019-05-14 19:54:06.91973), sending it over to the client, in the received response the milliseconds are gone (`2019-05-14T19:54:06Z`), sending this date back to the api and use it in a query like: `MyModel.query(on: req).filter(\.updatedAt > clientUpdatedAt).all()...` will not work as it should be, as the `clientUpdatedAt` will always be smaller, because `updatedAt` with milliseconds is of course always greater than `clientUpdatedAt` without milliseconds.

Right after you get a response:
```
let dataTask = session.dataTask(with: resourceURL) { data, _, _ in
    guard let jsonData = data else {
        ...
```
If you stop in the debugger and look at it:
`po String(data: jsonData, encoding: .utf8)` you can see that you received the iso representation without milliseconds.

### Steps to reproduce
Create a model with `createdAt`/`updatedAt` properties. Two handlers are enough, one `getAllHandler` which returns all entries, one createHandler which lets you create an entry. So create one and ask for it, you will receive the Date without milliseconds. So taking your fetched data, its date, how would you ask the api to give you only the results after/newer than your last `fetchedDate`? You can't you will always get the last result, you've already received, never an empty array.

I see two possibilities:
1. Define how Dates are saved in the database, with or without milliseconds. But you loose precision here.

2. When fetching data from the database, and creating a response e.g. like:
```
func getAllHandler(_ req: Request) throws -> Future<[MyModel]> {
    return MyModel.query(on: req).all()
  }
```
Add the `timeIntervalSinceReferenceDate : 579558135.977817` also to the payload, not just the iso representation, like it is done right now.

If you `po anyDateVariable` in the xcode debugger you get something like:
```
some : 2019-05-14 20:22:15 +0000
    - timeIntervalSinceReferenceDate : 579558135.977817
```
Which is weird, if you ask me. Now you see the problem, you can convert the `timeIntervalSinceReferenceDate` value to the upper iso date representation, but not the other way around, as this would result in `579558135.0` -> milliseconds lost.

### Environment

* Vapor Framework version: 3.3.0
* Vapor Toolbox version: 3.1.10
* OS version: 10.14.4

**Comment by nickpodratz** [comment]
I stumbled upon the same issue. The following `Date` extension helped me resolve my issue.
This is more like a workaround, though, as you have to remember applying this transformation. 
```
extension Date {

    /// Removes the date's milliseconds. The resulting date is temporally earlier or coincident to the original date.
    var trimmedToSeconds: Date {
        let timeIntervalWithoutMilliseconds = self.timeIntervalSince1970.rounded(.down)
        return Date(timeIntervalSince1970: timeIntervalWithoutMilliseconds)
    }
}
```

**Comment by leonidas-o** [comment]
If you don't need the fraction, of course, but if you need the precision than this won't work out. I've added a property of type `Timeinterval` to the model I'm sending in the api response, this way I'm not modifying any data and I'm not loosing any precision.

**Comment by nickpodratz** [comment]
Okay makes sense. I don't really need it. Do you know by chance how fluent does the rounding?
I tried using the `ISO8601DateFormatter` to convert the date into text and back, and also to apply the `rounded(rule:)` method with several rounding options - still my tests fail occasionally.
Any idea?

**Comment by leonidas-o** [comment]
I don't think fluent is doing any extra rounding nor any other modifications on user data by itself. It just uses ISO8601.
Actually it's at the beginning of my posted issue:
`The iso8601 representation has no milliseconds, so it is 2019-05-14T19:54:06Z. Therefore as soon as your api response contains dates, you loose the milliseconds information.`

**Comment by nickpodratz** [comment]
Yes I read that. But still, I cannot deterministically trim the original date in the same way fluent does. Like the example illustrates, it does not always round down.

**Comment by grundoon** [comment]
So I've been using this on Catalina/Ubuntu 18.04 with some success. It's not perfect because there's still rounding that can bite you right in the ms vicinity of the `\.createdAt` , but at least it's, well... a 1000x 😜  improvement.

(IIRC, 10.13  because that's when `.withFractionalSeconds` was introduced as an option)

`configure.swift`:

```swift
// use iso8601ms for dates
var contentConfig = ContentConfig.default()
let jsonEncoder = JSONEncoder()
let jsonDecoder = JSONDecoder()
if #available(OSX 10.13, *) {
    jsonEncoder.dateEncodingStrategy = .iso8601ms
    jsonDecoder.dateDecodingStrategy = .iso8601ms
} else {
    // Fallback on earlier versions
}
contentConfig.use(encoder: jsonEncoder, for: .json)
contentConfig.use(decoder: jsonDecoder, for: .json)
services.register(contentConfig)
```

```swift
import Foundation

@available(OSX 10.13, *)
extension ISO8601DateFormatter {
    convenience init(_ formatOptions: Options, timeZone: TimeZone? = TimeZone(secondsFromGMT: 0)) {
        self.init()
        self.formatOptions = formatOptions
        self.timeZone = timeZone
    }
}

@available(OSX 10.13, *)
extension Formatter {
    static let iso8601ms = ISO8601DateFormatter([.withInternetDateTime, .withFractionalSeconds])
}

@available(OSX 10.13, *)
extension Date {
    var iso8601ms: String {
        return Formatter.iso8601ms.string(from: self)
    }
}

@available(OSX 10.13, *)
extension String {
    var iso8601ms: Date? {
        return Formatter.iso8601ms.date(from: self)
    }
}

@available(OSX 10.13, *)
extension JSONDecoder.DateDecodingStrategy {
    static let iso8601ms = custom {
        let container = try $0.singleValueContainer()
        let string = try container.decode(String.self)
        guard let date = Formatter.iso8601ms.date(from: string) else {
            throw DecodingError.dataCorruptedError(
                in: container,
                debugDescription: "invalid format: " + string)
        }
        return date
    }
}

@available(OSX 10.13, *)
extension JSONEncoder.DateEncodingStrategy {
    static let iso8601ms = custom {
        var container = $1.singleValueContainer()
        try container.encode(Formatter.iso8601ms.string(from: $0))
    }
}
```

**Comment by tanner0101** [comment]
@grundoon's solution seems like the best solution here. Closing here since this is expected `JSONEncoder` / Foundation behavior.
